tests/scripts/branch-history

```diff
@@ -0,0 +1,209 @@
+
+   The option -fast, allows to skip the checks that require checkout of the
+   individual commits, which could be convenient when there are local
+   changes. Currently, only the first two file-level checks will be skipped.
+
+   To be useful, the script should be run in a clean repository that doesn't
+   have any uncommitted changes and is not in a "detached HEAD" state.
+
+   Example usage: $0 ../..
+                  $0 ../../ -commit 8d50f48
+                  $0 ../../ -range eb2017b..5dd5511
+                  $0 ../../ -range master..d5f6f7e
+                  $0 ../../ -commit d5f6f7e -fast > /dev/null
+
+EOF
+      exit
+      ;;
+esac
+
+# Maximum number of acceptable commits in the branch
+branch_max_commits=100
+# Maximum number of acceptable files changed in any commit in the branch
+commit_max_files_changed=100
+# Maximum number of acceptable lines changed in any commit in the branch
+commit_max_lines_changed=3000
+# Maximum acceptable size (in K) of any file changed in any commit in the branch
+file_max_size=200
+# Maximum number of acceptable lines added to any file changed in any commit in the branch
+file_max_lines_added=1000
+# Maximum number of acceptable lines deleted from any file changed in any commit in the branch
+file_max_lines_deleted=1000
+
+# Setup
+cur_dir=`pwd`
+mfem_dir=$1
+if [ "$mfem_dir" = "" ]; then
+    mfem_dir="../.."
+fi
+cd $mfem_dir
+mfem_dir=`pwd`
+shift
+
+current_branch="$(git symbolic-ref --short HEAD)"
+commit_range="master..${current_branch}"
+if [ "$1" = "-commit" ]; then
+    shift
+    # check just the specified commit
+    commit="$1"
+    commit_range="`git rev-parse ${commit}^`..${commit}"
+    shift
+fi
+if [ "$1" = "-range" ]; then
+    shift
+    commit_range="$1"
+    shift
+fi
+
+checkout_commits=true
+if [ "$1" = "-fast" ]; then
+    checkout_commits=false
+fi
+
+# Check if the branch contains unusually large number of commits
+num_commits="$(git rev-list --count --no-merges ${commit_range})"
+if [ ${num_commits} -gt ${branch_max_commits} ]
+then
+    printf '\033[35m'"This branch contains unusually many commits: " 1>&2
+    printf "${num_commits} commits not in master"'\033[0m\n' 1>&2
+fi
+
+# Check if the branch adds/modifies unusually large files (this can catch large
+# binary files not addressed by the checks below)
+while read -r obj; do
+    obj_fname="$(echo ${obj} | awk '{print $2}')"
+    obj_size="$(echo ${obj} | awk '{print int($1/1024)}')"
+    if [ ${obj_size} -gt ${file_max_size} ]
+    then
+        printf '\033[35m'"This branch adds/modifies a large file: " 1>&2
+        printf "${obj_fname} (size ${obj_size}K)"'\033[0m\n' 1>&2
+    fi
+done < <(git rev-list --objects --no-merges ${commit_range} | git cat-file --batch-check='%(objecttype) %(objectsize) %(rest)' | sed -n 's/^blob //p' | awk '{if (max[$2]<$1) max[$2]=$1} END {for (f in max) print max[f], f}' | sort --numeric-sort --key=1)
+
```

Comment:
Added Veselin's test on "branch" level. It could be useful...

Reply:
Why is this on my phone

Reply:
Yeah, that line is kinda fun :)

Reply:
Agreed, and I learned some awk (dynamic creation of a map, nice).
For the sake of readability, I reformatted it, but changed nothing else.

**Note: this portion on code is sensitive to apparently insignificant changes**
- Quotes ["] in `done <<< "${git_objects}" are mandatory.
- Usage of a while loop on read is also necessary to loop on lines (a for loop would loop on "words".